### PR TITLE
Add More Coil and Tank Choices for Desuperheater

### PR DESCRIPTION
**Pull request by joseph-robertson**
Design doc: https://docs.google.com/document/d/1JAcrgVEwUq6ZWMZSd6h9Eo5ZYXlI4DMt-AsPUMJ9OJ4/edit#

Ability to have on `CoilWaterHeatingDesuperheater`:
* `WaterHeaterStratified` tank object
* `CoilCoolingWaterToAirHeatPumpEquationFit` and `CoilCoolingDXMultiSpeed` heating source types

https://github.com/NREL/EnergyPlus/pull/7251
https://github.com/NREL/EnergyPlus/pull/7271

- [x] Check forward translation
- [ ] Check simulations

### CODE REVIEW

**Comment by joseph-robertson** [comment]
This appears to be creating the desired idf. Problem now is getting the correct energyplus folder in the `core-build` folder, so that we can functionally test.

**Comment by jmarrec** [comment]
Rebased onto `everybody_merge_now` branch. Had to resolved a few conflicts by hand, mostly `Coil:DX:Multispeed` has a reference in E+ to `\reference DesuperHeatingWaterOnlySources` but this PR was using  `\reference DesuperHeatingCoilSources`.

File: openstudiocore/resources/model/OpenStudio.idd [hunk 1]
```diff
@@ -15011,6 +15011,7 @@ OS:Coil:Cooling:DX:MultiSpeed,
        \required-field
        \type alpha
        \reference CoolingCoilsDXMultiSpeed
+       \reference DesuperHeatingCoilSources
```
Comment (jmarrec) [inline openstudiocore/resources/model/OpenStudio.idd]: This one in Coil:Cooling:DX:MultiSpeed is actually DesuperHeatingWaterOnlySources

Comment (jmarrec) [reply openstudiocore/resources/model/OpenStudio.idd]: Addressed in https://github.com/NREL/OpenStudio/pull/3565/commits/32125caace966483c82b4f67ea322cb5e6e576c7

**Comment by jmarrec** [comment]
@joseph-robertson Is that the matching regression test? https://github.com/NREL/OpenStudio-resources/pull/71

**Comment by joseph-robertson** [comment]
@jmarrec Yes, it is.

**Comment by jmarrec** [comment]
I will review this one more carefully tomorrow as well as the regression test but it looks really good from a cursory look. I doubt I'll have many review comments I cannot address myself but just in case try to set aside 30min tomorrow if there are any.

File: openstudiocore/src/energyplus/Test/CoilWaterHeatingDesuperheater_GTest.cpp [hunk 2]
```diff
@@ -57,7 +59,9 @@ TEST_F(EnergyPlusFixture, ForwardTranslator_CoilWaterHeatingDesuperheater) {
   std::vector<HVACComponent> testCoils = {
     CoilCoolingDXSingleSpeed(m),
     CoilCoolingDXTwoSpeed(m),
-    CoilCoolingDXTwoStageWithHumidityControlMode(m)
+    CoilCoolingDXTwoStageWithHumidityControlMode(m),
+    CoilCoolingDXVariableSpeed(m),
+    CoilCoolingDXMultiSpeed(m)
```
Comment (jmarrec) [inline openstudiocore/src/energyplus/Test/CoilWaterHeatingDesuperheater_GTest.cpp]: ideally we should add a test for the other remaining objects that are non DX (I should have done that in the first place), I'll do that.

File: openstudiocore/src/model/CoilWaterHeatingDesuperheater_Impl.hpp [hunk 3]
```diff
@@ -40,6 +40,7 @@ class Schedule;
 class CurveBiquadratic;
 class HVACComponent;
 class WaterHeaterMixed;
+class WaterHeaterStratified;
```
Comment (jmarrec) [inline openstudiocore/src/model/CoilWaterHeatingDesuperheater_Impl.hpp]: I understand why you added `WaterHeaterStratified` here, but the `WaterHeaterMixed` forward declaration wasn't needed to begin with, so I will remove these two.

File: openstudiocore/src/model/test/CoilWaterHeatingDesuperheater_GTest.cpp [hunk 4]
```diff
@@ -112,6 +118,121 @@ TEST_F(ModelFixture, CoilWaterHeatingDesuperheater_Remove)
     EXPECT_FALSE(heatRejection.demandOutletModelObject());
 }
 
+TEST_F(ModelFixture, CoilWaterHeatingDesuperheater_Remove2)
+{
+    Model model;
+    ScheduleCompact schedule = ScheduleCompact(model);
+    CoilWaterHeatingDesuperheater testObject = CoilWaterHeatingDesuperheater(model, schedule);
+    RefrigerationCondenserAirCooled condenser = RefrigerationCondenserAirCooled(model);
```
Comment (jmarrec) [inline openstudiocore/src/model/test/CoilWaterHeatingDesuperheater_GTest.cpp]: Unused

Comment (jmarrec) [reply openstudiocore/src/model/test/CoilWaterHeatingDesuperheater_GTest.cpp]: I cleaned that up

File: openstudiocore/src/energyplus/Test/CoilWaterHeatingDesuperheater_GTest.cpp (hunk 2 above)
Comment (jmarrec) [reply openstudiocore/src/energyplus/Test/CoilWaterHeatingDesuperheater_GTest.cpp]: Done in https://github.com/NREL/OpenStudio/pull/3565/commits/41a4de08b57878d3c798be2474cc6fba64d422ef

**Comment by jmarrec** [comment]
Merging now into the integration branch, I'll check regression side with the rest

**Comment by jmarrec** [comment]
Fix #3647

**Pull request merged**